### PR TITLE
TST: Add tests for PyDMLabel and DisplayFormat

### LINKED ISSUE #256: Tests for Label Widget

### PR DESCRIPTION
**Pull request by hmbui**
After merge, close #256 and #247 .

### CODE REVIEW

File: pydm/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    def value(self):
+        return self._value
+
+    @pyqtSlot(int)
+    @pyqtSlot(float)
+    @pyqtSlot(str)
+    @pyqtSlot(ndarray)
+    def receiveValue(self, val):
+        """
+        :param val: The value from a PyDM widget
```
Comment (hhslepicka) [inline pydm/tests/conftest.py]: Fix the Docstrings for the Numpy style.

File: pydm/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    prec_signal = pyqtSignal(int)
+    upper_ctrl_limit_signal = pyqtSignal([float], [int])
+    lower_ctrl_limit_signal = pyqtSignal([float], [int])
+
+    def __init__(self):
+        super(ConnectionSignals, self).__init__()
+        self._value = None
+
+    @property
+    def value(self):
```
Comment (hhslepicka) [inline pydm/tests/conftest.py]: Add Docstrings.

File: pydm/tests/widgets/test_label.py [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+    pydm_label.showUnits = True
+    signals.unit_signal[str].connect(pydm_label.unitChanged)
+    signals.unit_signal[str].emit(unit_name)
+    signals.new_value_signal[type(value)].emit(value)
+
+    assert (pydm_label.value == value)
+    assert (pydm_label.text() == expected)
+    assert (pydm_label.displayFormat == display_format)
+
+    # Now, turn of showUnits
```
Comment (hhslepicka) [inline pydm/tests/widgets/test_label.py]: Typo of -> off

File: pydm/widgets/display_format.py [hunk 4]
```diff
@@ -12,13 +13,20 @@ class DisplayFormat(object):
 
 
 def parse_value_for_display(value, precision, display_format_type=DisplayFormat.Default, string_encoding="utf_8", widget=None):
+    try:
```
Comment (hhslepicka) [inline pydm/widgets/display_format.py]: Replace this try/except with:
```python
    widget_name = "" if widget is None else widget.objectName()
```

Comment (hmbui) [reply pydm/widgets/display_format.py]: type(widget) shows there's a pyqtWrapperType around the QWidget:
>{type 'PyQt5.QtCore.pyqtWrapperType}<class 'PyQt5.QtWidgets.QWidget'>

which doesn't have the objectName method. That's why I catch the AttributeException.

File: pydm/tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    """
+    Wraparound signal collection to work as a fixture for all unit tests.
+    This fixture has a function scope to ensure we have a fresh fixture, i.e. a new set of signals, for every unit
+    test run.
+
+    Returns
+    -------
+    A collection of signals to bind to slots.
+    """
+    return ConnectionSignals()
```
[the remark this answers is not in the document]
Comment (hmbui) [reply pydm/tests/conftest.py]: pytest doesn't support a class fixture, and I wanted to get an instance of ConnectionSignals().

File: pydm/widgets/display_format.py [hunk 6]
```diff
@@ -1,6 +1,13 @@
 import math
 import numpy as np
-import sys
+import tempfile
+import logging
+
+
+logger = logging.getLogger(__name__)
+_, file_path = tempfile.mkstemp(suffix=".log")
+handler = logging.FileHandler(file_path)
```
Comment (hhslepicka) [inline pydm/widgets/display_format.py]: Wrong place for the handler... maybe conftest.py?

File: pydm/tests/widgets/test_label.py [hunk 7]
```diff
@@ -540,8 +540,10 @@ def test_value_changed_incorrect_display_format(qtbot, signals, capfd, value, di
     signals.new_value_signal[type(value)].emit(value)
     pydm_label.displayFormat = display_format
 
-    out, err = capfd.readouterr()
-    assert expected in err
+    # Make sure logging capture the error, and have the correct error message
+    for record in caplog.records:
+        assert record.levelname == 'ERROR'
```
Comment (hhslepicka) [inline pydm/tests/widgets/test_label.py]: Wouldn't it be better to switch to `levelno` and compare with `logging.ERROR` instead?

File: pydm/widgets/display_format.py [hunk 8]
```diff
@@ -12,13 +13,20 @@ class DisplayFormat(object):
...
     if display_format_type == DisplayFormat.Default:
         return value
     elif display_format_type == DisplayFormat.String:
         if isinstance(value, np.ndarray):
             try:
                 r = value.tobytes().decode(string_encoding)
-                print("Could not decode {} using {} at widget named '{}'.".format(value, string_encoding, widget.objectName()))
+                sys.stderr.write("Could not decode\n")
+                print("Could not decode {0} using {1} at widget named '{2}'."
+                      .format(value, string_encoding, widget_name))
```
[the remark this answers is not in the document]
Comment (hmbui) [reply pydm/widgets/display_format.py]: Yes, we'll be using logging.error from now on.

File: pydm/tests/widgets/test_display_format.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        The numeric precision to consider during formatting
+    display_format : int
+        The format type for the provided value
+    widget : PyDWidget
+        The widget that will display the formatted value. This object can be None
+    expected : str
+        The expected formatted presentation of the provided value
+    """
+    parsed_value = parse_value_for_display(
+        value, precision, display_format_type=display_format, widget=widget)
```
[the remark this answers is not in the document]
Comment (hhslepicka) [reply pydm/tests/widgets/test_display_format.py]: The indent is right on this case according to PEP8...
From the link that you mentioned...
```
# Hanging indents should add a level.
foo = long_function_name(
    var_one, var_two,
    var_three, var_four)
```

And PyCharm with the PEP8 plugin enabled gives the same result as above.
What would be right to you in this case?

[the remark this answers is not in the document]
Comment (teddyrendahl) [reply pydm/tests/widgets/test_display_format.py]: Nevermind. My mistake!

**Comment by hhslepicka** [comment]
Merging the changes! Thank you @hmbui ! Great work here! 🎉

**Pull request merged**